File: tools/eaasp-common/src/eaasp_common/business_flow.py
"""EAASP Business Flow — vertical cross-layer business-context binding.

Per v3.15.1 design (PLATFORM_OBSERVABILITY_DESIGN.md §3.1 / §3.4). The
business flow is the **logical spine** of platform observability: a
single end-to-end business request binds together every cross-layer
event (LLM call, tool call, governance decision, memory write, SSE
chunk) under a stable ``(session_id, skill_id, business_object_id)``
key.

Two parallel IDs coexist on every event:

- ``trace_id`` (W3C) — solves "where in the call chain is the slow part".
- ``business_key`` — solves "where in the business flow did it break".

This module provides:

1. ``BusinessKey`` — frozen dataclass + validation.
2. ``BusinessFlowContext`` — contextvar-based per-task / per-request
   context. All EAASP Python tools can read the active business key
   without threading it through every function signature.
3. ``serialize_business_key`` / ``parse_business_key_header`` — wire
   format for the ``X-Business-Key`` HTTP/gRPC header.
4. ``require_business_key_from_request`` — FastAPI dependency that
   pulls the business key from the request headers and stashes it in
   the contextvar for downstream handlers.

The module is intentionally **dependency-free** (stdlib only) so every
EAASP tool (L2 / L3 / L4 / L5) can import it without taking on extra
deps. The HTTP/gRPC middleware lives in each tool (so the common
package stays transport-agnostic).

Wire format
-----------

The ``X-Business-Key`` header is a single line::

    <session_id>|<skill_id>|<business_object_id>

- ``session_id`` is required; empty header is rejected.
- ``skill_id`` and ``business_object_id`` are optional (empty string OK).
- Pipe (``|``) characters inside a field are not allowed; the parser
  rejects the header rather than silently truncating.

The choice of pipe-with-empty-allowed keeps the wire format
human-readable in debug output while still being parseable by a
single ``str.split('|', 2)`` call.
"""

from __future__ import annotations

import contextvars
from dataclasses import dataclass
from typing import Iterator

# Max length for any single field — keeps the header bounded and
# guards against pathological inputs (DoS via huge header).
_MAX_FIELD_LEN = 256

# ─── Domain model ───────────────────────────────────────────────────────────


@dataclass(frozen=True)
class BusinessKey:
    """Vertical cross-layer business-context binding key.

    All three fields are strings. ``session_id`` is mandatory (v3.15
    sets it as the primary join key for cross-table timeline queries);
    the other two are optional — at least one of them should be set
    for the key to be meaningful as a business identifier.

    Validation enforces non-empty ``session_id``, no pipe characters,
    and bounded length on every field. The validator raises ``ValueError``
    on violation (caller decides whether to log-and-drop, return HTTP
    400, etc.).
    """

    session_id: str
    skill_id: str = ""
    business_object_id: str = ""

    def __post_init__(self) -> None:
        if not self.session_id:
            raise ValueError("BusinessKey.session_id must be non-empty")
        for field_name in ("session_id", "skill_id", "business_object_id"):
            value = getattr(self, field_name)
            if not isinstance(value, str):
                raise ValueError(
                    f"BusinessKey.{field_name} must be a str, got {type(value).__name__}"
                )
            if len(value) > _MAX_FIELD_LEN:
                raise ValueError(
                    f"BusinessKey.{field_name} length {len(value)} exceeds max {_MAX_FIELD_LEN}"
                )
            if "|" in value:
                raise ValueError(
                    f"BusinessKey.{field_name} must not contain '|' (reserved as wire separator)"
                )

    # ─── Convenience predicates ────────────────────────────────────────────

    @property
    def is_meaningful(self) -> bool:
        """True iff at least one of skill_id / business_object_id is set.

        A key with only ``session_id`` still routes events to the same
        session timeline but doesn't bind to a business object — useful
        for "infrastructure" flows (health checks, etc.).
        """
        return bool(self.skill_id or self.business_object_id)

    def matches(self, other: "BusinessKey") -> bool:
        """Return True iff both keys refer to the same business flow.

        Matching is prefix-based: a key with empty skill_id matches
        any other key with the same session_id. This lets a session
        start without a skill and acquire one later without breaking
        the timeline query.
        """
        if self.session_id != other.session_id:
            return False
        if self.skill_id and other.skill_id and self.skill_id != other.skill_id:
            return False
        if (
            self.business_object_id
            and other.business_object_id
            and self.business_object_id != other.business_object_id
        ):
            return False
        return True

    def to_header(self) -> str:
        """Render as the wire format used in ``X-Business-Key`` headers.

        Always emits three pipe-separated fields; empty fields are
        emitted as empty strings. Round-trips through
        ``parse_business_key_header``.
        """
        return f"{self.session_id}|{self.skill_id}|{self.business_object_id}"

    def to_dict(self) -> dict[str, str]:
        return {
            "session_id": self.session_id,
            "skill_id": self.skill_id,
            "business_object_id": self.business_object_id,
        }

    def __str__(self) -> str:
        return self.to_header()


# ─── Wire format ────────────────────────────────────────────────────────────


def parse_business_key_header(raw: str | None) -> BusinessKey | None:
    """Parse a wire-format ``X-Business-Key`` header.

    Returns None when:
    - ``raw`` is None or empty (header missing — caller treats as "not
      part of any business flow"; per v3.15.1 design, business_key is
      optional).

    Raises ``ValueError`` when the header is malformed (wrong number
    of fields, invalid characters, length > max). Callers should log
    + drop the header (do not raise to the user) — bad headers are
    treated as missing keys for resilience.
    """
    if raw is None or raw == "":
        return None
    parts = raw.split("|", 2)
    if len(parts) != 3:
        raise ValueError(
            f"X-Business-Key must have 3 pipe-separated fields, got {len(parts)}"
        )
    return BusinessKey(
        session_id=parts[0],
        skill_id=parts[1],
        business_object_id=parts[2],
    )


def serialize_business_key(key: BusinessKey) -> str:
    """Render a ``BusinessKey`` to wire format. Inverse of ``parse_*``."""
    return key.to_header()


# ─── Context propagation ────────────────────────────────────────────────────
#
# A single contextvar holds the *active* business key for the current
# asyncio task. Middleware sets it on request entry and resets it on
# exit. Library code can read it via ``get_current_business_key()``
# without needing the key passed in every function signature.

_business_key_var: contextvars.ContextVar[BusinessKey | None] = contextvars.ContextVar(
    "eaasp_business_key", default=None
)


def set_current_business_key(key: BusinessKey | None) -> contextvars.Token:
    """Set the active business key. Returns the token for ``reset()``."""
    return _business_key_var.set(key)


def reset_current_business_key(token: contextvars.Token) -> None:
    """Restore the prior business key. Call in a ``finally`` after middleware."""
    _business_key_var.reset(token)


def get_current_business_key() -> BusinessKey | None:
    """Return the active business key, or None if none is bound."""
    return _business_key_var.get()


def require_current_business_key() -> BusinessKey:
    """Return the active business key or raise ``LookupError``.

    Use in code paths that REQUIRE a business key to function (e.g.
    cross-layer event tagging). Library code that should still work
    for non-business-flow events uses ``get_current_business_key()``.
    """
    key = _business_key_var.get()
    if key is None:
        raise LookupError("no active business key (set via set_current_business_key)")
    return key


# ─── Helper context manager ────────────────────────────────────────────────


class business_key_scope:
    """Context manager that binds a ``BusinessKey`` for the duration of a block.

    Usage::

        with business_key_scope(my_key):
            ...  # downstream code can call get_current_business_key()
    """

    def __init__(self, key: BusinessKey | None) -> None:
        self._key = key
        self._token: contextvars.Token | None = None

    def __enter__(self) -> "business_key_scope":
        self._token = set_current_business_key(self._key)
        return self

    def __exit__(self, *exc: object) -> None:
        del exc  # unused; context-manager protocol requires this signature
        if self._token is not None:
            reset_current_business_key(self._token)
            self._token = None

    # Make the class async-context-manager compatible too — the
    # asynccontextmanager decorator wraps this class when needed.
    async def __aenter__(self) -> "business_key_scope":
        return self.__enter__()

    async def __aexit__(self, *exc: object) -> None:
        self.__exit__(*exc)


def iter_business_key_contexts() -> Iterator[BusinessKey | None]:
    """Yield the active business key once. Helper for tests."""
    yield get_current_business_key()

File: crates/grid-runtime/src/business_flow.rs
//! EAASP Business Flow — vertical cross-layer business-context binding
//! for the L1 Rust runtime.
//!
//! Per OBSTACK_DESIGN.md §3.1 / §3.4. Mirror of
//! ``tools/eaasp-common/src/eaasp_common/business_flow.py`` — same
//! wire format, same field semantics, same per-task propagation rule.
//!
//! Differences from the Python module:
//! - Storage: `task_local!` (tokio) instead of Python's `contextvars`.
//!   tokio's `task_local` provides per-async-task isolation, which is
//!   the Rust equivalent of Python's contextvar for request-scoped
//!   context propagation. The accessor functions stay the same shape.
//! - Header parser: `X-Business-Key` is a single ASCII line. We parse
//!   it lazily from tonic `Metadata` extensions on the gRPC path.
//! - Error: Python `ValueError` → Rust `BusinessKeyError::Parse`; the
//!   upper layer maps that to gRPC `INVALID_ARGUMENT`.
//!
//! Wire format (identical to the Python module):
//!
//! ```text
//! <session_id>|<skill_id>|<business_object_id>
//! ```
//!
//! - `session_id` is required (non-empty).
//! - `skill_id` and `business_object_id` are optional (empty OK).
//! - Pipe (`|`) characters inside a field are rejected.

use std::cell::Cell;
use thiserror::Error;

pub const MAX_FIELD_LEN: usize = 256;

// ─── Domain model ───────────────────────────────────────────────────────────

/// Three-tuple that vertically binds one end-to-end business request
/// across L0–L5. Mirror of ``eaasp_common.business_flow.BusinessKey``.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct BusinessKey {
    pub session_id: String,
    pub skill_id: String,
    pub business_object_id: String,
}

#[derive(Debug, Error, PartialEq, Eq)]
pub enum BusinessKeyError {
    #[error("BusinessKey.session_id must be non-empty")]
    EmptySessionId,
    #[error("BusinessKey.{0} must be a String, got {1}")]
    NotAString(&'static str, String),
    #[error("BusinessKey.{0} length {1} exceeds max {2}")]
    TooLong(&'static str, usize, usize),
    #[error("BusinessKey.{0} must not contain '|' (reserved as wire separator)")]
    PipeInField(&'static str),
    #[error(
        "X-Business-Key must have 3 pipe-separated fields, got {0}"
    )]
    WrongFieldCount(usize),
}

impl BusinessKey {
    pub fn new(
        session_id: impl Into<String>,
        skill_id: impl Into<String>,
        business_object_id: impl Into<String>,
    ) -> Result<Self, BusinessKeyError> {
        let session_id = session_id.into();
        let skill_id = skill_id.into();
        let business_object_id = business_object_id.into();
        let key = Self {
            session_id,
            skill_id,
            business_object_id,
        };
        key.validate()?;
        Ok(key)
    }

    pub fn validate(&self) -> Result<(), BusinessKeyError> {
        validate_field("session_id", &self.session_id)?;
        validate_field("skill_id", &self.skill_id)?;
        validate_field(
            "business_object_id",
            &self.business_object_id,
        )?;
        Ok(())
    }

    /// True iff at least one of skill_id / business_object_id is set.
    pub fn is_meaningful(&self) -> bool {
        !self.skill_id.is_empty() || !self.business_object_id.is_empty()
    }

    /// Prefix-based matching (same semantics as Python's
    /// ``BusinessKey.matches``): non-empty fields must match exactly;
    /// empty fields are treated as wildcards. Lets a session start
    /// without a skill and acquire one later without breaking the
    /// timeline query.
    pub fn matches(&self, other: &BusinessKey) -> bool {
        if self.session_id != other.session_id {
            return false;
        }
        if !self.skill_id.is_empty()
            && !other.skill_id.is_empty()
            && self.skill_id != other.skill_id
        {
            return false;
        }
        if !self.business_object_id.is_empty()
            && !other.business_object_id.is_empty()
            && self.business_object_id != other.business_object_id
        {
            return false;
        }
        true
    }

    /// Render as the wire format used in ``X-Business-Key`` headers.
    pub fn to_header(&self) -> String {
        format!(
            "{}|{}|{}",
            self.session_id, self.skill_id, self.business_object_id
        )
    }

    /// Parse a wire-format ``X-Business-Key`` header.
    ///
    /// Returns `Ok(None)` when the header is missing or empty — this is
    /// the "not part of any business flow" case (per OBSTACK §3.8 #1,
    /// business_key is optional during migration).
    pub fn parse_header(raw: Option<&str>) -> Result<Option<Self>, BusinessKeyError> {
        match raw {
            None => Ok(None),
            Some("") => Ok(None),
            Some(s) => {
                let parts: Vec<&str> = s.splitn(3, '|').collect();
                if parts.len() != 3 {
                    return Err(BusinessKeyError::WrongFieldCount(parts.len()));
                }
                Self::new(parts[0], parts[1], parts[2]).map(Some)
            }
        }
    }
}

fn validate_field(name: &'static str, value: &str) -> Result<(), BusinessKeyError> {
    if name == "session_id" && value.is_empty() {
        return Err(BusinessKeyError::EmptySessionId);
    }
    if value.len() > MAX_FIELD_LEN {
        return Err(BusinessKeyError::TooLong(name, value.len(), MAX_FIELD_LEN));
    }
    if value.contains('|') {
        return Err(BusinessKeyError::PipeInField(name));
    }
    Ok(())
}

impl std::fmt::Display for BusinessKey {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.to_header())
    }
}

impl std::str::FromStr for BusinessKey {
    type Err = BusinessKeyError;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match Self::parse_header(Some(s))? {
            Some(k) => Ok(k),
            None => Err(BusinessKeyError::EmptySessionId),
        }
    }
}

// ─── Per-task context propagation ──────────────────────────────────────────
//
// tokio's `task_local!` provides per-async-task storage. Each gRPC
// request runs in its own task, so setting the business key on entry
// and resetting it on exit gives us the same isolation guarantees as
// Python's `contextvars`. The default is `None` (no business flow).

tokio::task_local! {
    static BUSINESS_KEY: Cell<Option<BusinessKey>>;
}

/// Set the active business key for the current async task. Returns a
/// `TaskLocalFuture` that, when dropped, restores the prior value.
///
/// Mirror of Python's ``set_current_business_key`` + ``reset_*`` pair.
pub fn scope<F>(key: Option<BusinessKey>, fut: F) -> impl std::future::Future<Output = F::Output>
where
    F: std::future::Future,
{
    BUSINESS_KEY.scope(Cell::new(key), fut)
}

/// Read the active business key inside the scoped future.
///
/// Use `require()` instead when the key is mandatory for the calling
/// code path.
pub fn current() -> Option<BusinessKey> {
    BUSINESS_KEY
        .try_with(|c| c.take())
        .ok()
        .flatten()
}

/// Like `current()` but returns `BusinessKeyError::EmptySessionId`
/// when no key is bound. The error name is reused (`EmptySessionId`)
/// for symmetry with the constructor; callers that want a distinct
/// "missing key" signal should map accordingly.
pub fn require() -> Result<BusinessKey, BusinessKeyError> {
    current().ok_or(BusinessKeyError::EmptySessionId)
}

// ─── Tests ──────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    fn sample() -> BusinessKey {
        BusinessKey::new("sess-A", "skill-thr", "Transformer-001")
            .expect("sample key is valid")
    }

    #[test]
    fn header_round_trip() {
        let k = sample();
        let header = k.to_header();
        let parsed = BusinessKey::parse_header(Some(&header))
            .expect("parse ok")
            .expect("non-None key");
        assert_eq!(parsed, k);
    }

    #[test]
    fn parse_header_missing_returns_none() {
        assert_eq!(BusinessKey::parse_header(None).unwrap(), None);
        assert_eq!(BusinessKey::parse_header(Some("")).unwrap(), None);
    }

    #[test]
    fn parse_header_rejects_too_few_fields() {
        let err = BusinessKey::parse_header(Some("only_one_field")).unwrap_err();
        assert_eq!(err, BusinessKeyError::WrongFieldCount(1));

        let err = BusinessKey::parse_header(Some("a|b")).unwrap_err();
        assert_eq!(err, BusinessKeyError::WrongFieldCount(2));
    }

    #[test]
    fn parse_header_rejects_pipe_in_field() {
        // `splitn(3, '|')` splits "a|b|c|d" into ["a", "b", "c|d"]; the
        // third field still contains a pipe, so validation must reject
        // it as PipeInField rather than silently accepting 3 fields.
        let err = BusinessKey::parse_header(Some("a|b|c|d")).unwrap_err();
        assert_eq!(err, BusinessKeyError::PipeInField("business_object_id"));

        // Pipe in field 0 (session_id) is caught directly.
        let err = BusinessKey::parse_header(Some("a|b|with|pipe"))
            .unwrap_err();
        // Splitn gives ["a", "b", "with|pipe"]; business_object_id has pipe.
        assert_eq!(err, BusinessKeyError::PipeInField("business_object_id"));
    }

    #[test]
    fn validate_rejects_empty_session_id() {
        assert_eq!(
            BusinessKey::new("", "skill", "obj").unwrap_err(),
            BusinessKeyError::EmptySessionId
        );
    }

    #[test]
    fn validate_rejects_oversize_field() {
        let big = "x".repeat(MAX_FIELD_LEN + 1);
        let err = BusinessKey::new("sess", big.clone(), "").unwrap_err();
        assert_eq!(err, BusinessKeyError::TooLong("skill_id", MAX_FIELD_LEN + 1, MAX_FIELD_LEN));
    }

    #[test]
    fn matches_is_prefix_based() {
        let a = BusinessKey::new("sess-A", "skill-thr", "Transformer-001").unwrap();
        let b = BusinessKey::new("sess-A", "skill-thr", "Transformer-001").unwrap();
        assert!(a.matches(&b));

        let c = BusinessKey::new("sess-A", "", "").unwrap(); // session-only
        assert!(c.matches(&a)); // session-only matches anything in same session

        let d = BusinessKey::new("sess-B", "skill-thr", "Transformer-001").unwrap();
        assert!(!a.matches(&d)); // different session

        let e = BusinessKey::new("sess-A", "skill-other", "Transformer-001").unwrap();
        assert!(!a.matches(&e)); // different skill
    }

    #[test]
    fn is_meaningful_requires_skill_or_object() {
        let session_only = BusinessKey::new("sess", "", "").unwrap();
        assert!(!session_only.is_meaningful());

        let with_skill = BusinessKey::new("sess", "skill", "").unwrap();
        assert!(with_skill.is_meaningful());

        let with_object = BusinessKey::new("sess", "", "obj").unwrap();
        assert!(with_object.is_meaningful());
    }

    #[tokio::test]
    async fn task_local_scope_isolates_keys() {
        let k = sample();

        // Inside scope: current key is `k`.
        scope(Some(k.clone()), async {
            let cur = current().expect("key set");
            assert_eq!(cur, k);

            // Nested scope with different key shadows outer (per task,
            // not per scope — this is the same task).
            let inner = BusinessKey::new("sess-B", "skill-other", "obj").unwrap();
            scope(Some(inner.clone()), async {
                let cur = current().expect("nested key set");
                assert_eq!(cur, inner);
            })
            .await;
        })
        .await;

        // After outer scope exits, no key is bound in this task.
        assert_eq!(current(), None);

        // Different task: independent state (per-task isolation is
        // the Rust equivalent of Python's contextvar).
        let k2 = sample();
        tokio::spawn(async move {
            scope(Some(k2.clone()), async {
                let cur = current().expect("key in spawned task");
                assert_eq!(cur, k2);
            })
            .await;
        })
        .await
        .expect("spawned task completed");
    }

    #[tokio::test]
    async fn require_returns_error_outside_scope() {
        // Outside any scope — current() is None, require() errors.
        assert_eq!(current(), None);
        let err = require().unwrap_err();
        assert_eq!(err, BusinessKeyError::EmptySessionId);
    }
}
